sdk/dotnet/Docker.cs

```diff
@@ -0,0 +1,644 @@
+                if (code > 0)
+                {
+                    continue;
+                }
+
+                cacheFromImages.Add(image);
+            }
+
+            return cacheFromImages.ToArray();
+        }
+
+        private static async Task<(string imageId, string[] stages)> BuildImageAsync(
+            string imageName, Union<string, DockerBuildUnwrap> pathOrBuild, Resource logResource, string[]? cacheFrom)
+        {
+            DockerBuildUnwrap build = pathOrBuild.IsT0 ? new DockerBuildUnwrap { Context = pathOrBuild.AsT0 } : pathOrBuild.AsT1;
+
+            // If the build context is missing, default it to the working directory.
+            if (build.Context == null)
+            {
+                build.Context = ".";
+            }
+
+            Log.Info(
+                $"Building container image '{imageName}': context={build.Context}" +
+                (build.Dockerfile != null ? $", dockerfile={build.Dockerfile}" : "") +
+                (build.Args != null ? $", args={JsonSerializer.Serialize(build.Args)}" : ""),
+                logResource, ephemeral: true);
+
+            // If the container build specified build stages to cache, build each in turn.
+            var stages = new List<string>();
+            if (build.CacheFrom?.Stages != null)
+            {
+                foreach (var stage in build.CacheFrom.Stages)
+                {
+                    await RunDockerBuild(LocalStageImageName(imageName, stage), build, cacheFrom, logResource, stage);
+                    stages.Add(stage);
+                }
+            }
+
+            //// Invoke Docker CLI commands to build.
+            await RunDockerBuild(imageName, build, cacheFrom, logResource);
+
+            // Finally, inspect the image so we can return the SHA digest. Do not forward the output of this
+            // command this to the CLI to show the user.
+            var inspectResult = await RunCommandThatMustSucceed(
+                "docker", new[] { "image", "inspect", "-f", "{{.Id}}", imageName }, logResource);
+            if (string.IsNullOrEmpty(inspectResult))
+            {
+                throw new ResourceException($"No digest available for image {imageName}", logResource);
+            }
+
+            // From https://docs.docker.com/registry/spec/api/#content-digests
+            //
+            // the image id will be a "algorithm:hex" pair.  We don't care about the algorithm part.  All we
+            // want is the unique portion we can use elsewhere.  Since we are also going to place this in an
+            // image tag, we also don't want the colon, as that's not legal there.  So simply grab the hex
+            // portion after the colon and return that.
+
+            var imageId = inspectResult.Trim();
+            var colonIndex = imageId.LastIndexOf(":");
+            imageId = colonIndex < 0 ? imageId : imageId.Substring(colonIndex + 1);
+
+            return (imageId, stages.ToArray());
```

Comment:
note: i'm skimming most of this code with the assumption you're copying the js sdk.